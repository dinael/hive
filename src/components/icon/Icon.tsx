import { FC } from 'react'
import styles from './Icon.module.scss'
import sprite from '@/assets/sprite.svg'

export type IconProps = {
  name: string
  size?: 'default' | 'xs' | 's' | 'm' | 'l' | 'xl'
  title?: string
  viewBox?: number
  className?: string
}

export const Icon: FC<IconProps> = ({
  name = 'info',
  size  = 'default',
  title = '',
  viewBox = 24,
  className = '',
  ...props
}: IconProps) => {

  const iconViewBox = `0 0 ${viewBox} ${viewBox}`
  const iconTitle = title || name
  const iconName = `${sprite}#${name}`

  const iconStyles = `
    ${styles['icon-component']}
    ${styles[size ? size : 'm']}
    ${className || ''}
  `
  return (
    <span
      aria-hidden="true"
      className={iconStyles}
      {...props}>
      <svg viewBox={iconViewBox}>
        <title>{ iconTitle }</title>
        <use xlinkHref={iconName} />
      </svg>
    </span>
  )
}

export default Icon
