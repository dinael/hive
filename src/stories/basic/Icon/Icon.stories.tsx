import { Icon, IconProps } from '@/components/icon/Icon'

export default {
  title: 'Components/Basic/Icon',
  component: Icon,
  parameters: {
    vitest: {
      testFile: "Icon.test.tsx"
    }
  },
  tags: ['autodocs'],
  argTypes: {
    name: {
      name: 'Name',
      control: 'text'
    },
    size: {
      name: 'Size',
      control: 'text'
    },
    title: {
      name: 'Title',
      control: 'text'
    },
    viewBox: {
      name: 'ViewBox',
      control: 'number'
    },
  }
}

export const Default = (args: IconProps) => <Icon {...args} />

Default.args = {
  name: 'close',
  size : 'm',
  title: '',
  viewBox: 24,
}

Default.parameters = {
  design: [
    {
      type: 'link',
      name: 'Icon source',
      label: 'Icon source: mingcute.com',
      target: '_blank',
      url: 'https://www.mingcute.com/',
    },
    {
      type: 'iframe',
      name: 'Icon use example',
      url: 'https://codepen.io/dinael/full/vYmLVpa',
    }
  ]
}

