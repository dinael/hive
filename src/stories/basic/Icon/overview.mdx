import { Canvas, Meta, Title, IconGallery, IconItem } from '@storybook/blocks'
import { Figma } from 'storybook-addon-designs/blocks';
import Iconography from '@/stories/tokens/iconography.mdx'

import { Icon as IconExample } from '@/components/icon/Icon'

<Meta title="Components/Basic/Icon/Overview" />
# Icon component
The Icon component makes it easy to efficiently and accessibly incorporate icons into your user interface.

To use the Icon component, first ensure you have an **SVG sprite** file containing symbols for your icons. An icon is a visual representation of an object and can be used to represent actions or objects in the interface. Icons are created using **SVG**.

**Note:** \
Ensure that the **IDs** of the symbols in your **SVG sprite** file match the names of the icons you intend to use with the Icon component.

## How create a SVG sprite
```html:
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" fill="currentColor" height="0" width="0">
  <symbol id="chevron-up" viewBox="0 0 24 24">
    <path d="..."/>
  </symbol>
  <symbol id="chevron-right" viewBox="0 0 24 24">
    <path d="..."/>
  </symbol>
<svg/>
```
## Usage
```
<Icon name="icon-name" size="m" title="Optional description" />
```

<Iconography />

## Changelog
### v0.0.11 (Wed May 31 2023)

## Figma source
<Figma
  height="30%"
  collapsable
  url="https://www.figma.com/file/Cx0sZBqrc7kGW72bcmy90S/tokens?type=design&node-id=3-27&t=ThF1FKZYYApN4drF-4"
/>
