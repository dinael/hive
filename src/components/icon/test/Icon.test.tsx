import { describe, test, expect } from 'vitest'
import { render } from '@testing-library/react'
import { Icon, IconProps } from '../Icon'

describe('Icon', () => {
  const defaultProps: IconProps = {
    name: 'close',
    size : 'xl',
    color : 'currentColor',
    title: 'This is a title for a test',
    viewBox: 24,
    className: 'className-test'
  }

  test('Icon renders snapshot', () => {
    const { container } = render(<Icon {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('Icon renders correctly', () => {
    const { container } = render(<Icon {...defaultProps} />)
    const iconElement = container.querySelector('span')
    expect(iconElement?.className).toContain('icon-component')
    expect(iconElement?.tagName).toBe('SPAN')
    expect(container).toMatchSnapshot()
  })

  test('Icon renders correctly viewBox', () => {
    const { container } = render(<Icon {...defaultProps} />)
    const iconElement = container.querySelector('span > svg')
    expect(iconElement?.getAttribute('viewBox')).toBe('0 0 24 24')
    expect(container).toMatchSnapshot()
  })

  test('Icon renders correctly svg title', () => {
    const { container } = render(<Icon {...defaultProps} />)
    const iconElement = container.querySelector('span > svg > title')
    expect(iconElement?.textContent).toBe(defaultProps.title)
    expect(container).toMatchSnapshot()
  })

  test('Icon renders with a custom class', () => {
    const { container } = render(<Icon {...defaultProps} />)
    const blockquoteElement = container.querySelector('span')
    expect(blockquoteElement?.className).toContain(defaultProps.className)
    expect(container).toMatchSnapshot()
  })
})
