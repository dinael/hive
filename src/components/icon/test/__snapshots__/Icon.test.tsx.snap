// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Icon > Icon renders correctly 1`] = `
<div>
  <span
    aria-hidden="true"
    class="
    icon-component
    xl
    className-test
  "
  >
    <svg
      viewBox="0 0 24 24"
    >
      <title>
        This is a title for a test
      </title>
      <use
        xlink:href="/src/assets/sprite.svg#close"
      />
    </svg>
  </span>
</div>
`;

exports[`Icon > Icon renders correctly svg title 1`] = `
<div>
  <span
    aria-hidden="true"
    class="
    icon-component
    xl
    className-test
  "
  >
    <svg
      viewBox="0 0 24 24"
    >
      <title>
        This is a title for a test
      </title>
      <use
        xlink:href="/src/assets/sprite.svg#close"
      />
    </svg>
  </span>
</div>
`;

exports[`Icon > Icon renders correctly viewBox 1`] = `
<div>
  <span
    aria-hidden="true"
    class="
    icon-component
    xl
    className-test
  "
  >
    <svg
      viewBox="0 0 24 24"
    >
      <title>
        This is a title for a test
      </title>
      <use
        xlink:href="/src/assets/sprite.svg#close"
      />
    </svg>
  </span>
</div>
`;

exports[`Icon > Icon renders snapshot 1`] = `
<div>
  <span
    aria-hidden="true"
    class="
    icon-component
    xl
    className-test
  "
  >
    <svg
      viewBox="0 0 24 24"
    >
      <title>
        This is a title for a test
      </title>
      <use
        xlink:href="/src/assets/sprite.svg#close"
      />
    </svg>
  </span>
</div>
`;

exports[`Icon > Icon renders with a custom class 1`] = `
<div>
  <span
    aria-hidden="true"
    class="
    icon-component
    xl
    className-test
  "
  >
    <svg
      viewBox="0 0 24 24"
    >
      <title>
        This is a title for a test
      </title>
      <use
        xlink:href="/src/assets/sprite.svg#close"
      />
    </svg>
  </span>
</div>
`;
